WalletWasabi/Wabisabi/CredentialIssuer.cs

```diff
@@ -0,0 +1,176 @@
+			CoordinatorParameters = CoordinatorSecretKey.ComputeCoordinatorParameters();
+			RandomNumberGenerator = Guard.NotNull(nameof(randomNumberGenerator), randomNumberGenerator);
+		}
+
+		// Keeps track of the used serial numbers. This is part of 
+		// the double-spending prevention mechanism.
+		private HashSet<GroupElement> SerialNumbers { get; } = new HashSet<GroupElement>();
+		
+		// Canary test check to ensure credential balance is never negative
+		private Money Balance { get; set; } = Money.Zero;
+
+		private WasabiRandom RandomNumberGenerator { get; }
+
+		private CoordinatorSecretKey CoordinatorSecretKey { get; }
+		
+		private CoordinatorParameters CoordinatorParameters { get; }
+
+		// Gets the number of credentials that have to be requested/presented
+		// This parameter is called `k` in the wabisabi paper.
+		public int NumberOfCredentials { get; }
+
+		public RegistrationResponse HandleRequest(RegistrationRequest registrationRequest)
+		{
+			Guard.NotNull(nameof(registrationRequest), registrationRequest);
+
+			var requested = registrationRequest.Requested ?? Enumerable.Empty<IssuanceRequest>();
+			var presented = registrationRequest.Presented ?? Enumerable.Empty<CredentialPresentation>();
+
+			var requestedCount = requested.Count();  
+			if (requestedCount != NumberOfCredentials)
+			{
+				throw new WabiSabiException(
+					WabiSabiErrorCode.InvalidNumberOfRequestedCredentials, 
+					$"{NumberOfCredentials} credential requests were expected but {requestedCount} were received.");
+			}
+
+			var presentedCount = presented.Count();
+			var requiredNumberOfPresentations = registrationRequest.IsNullRequest ? 0 : NumberOfCredentials;
+			if (presentedCount != requiredNumberOfPresentations)
+			{
+				throw new WabiSabiException(
+					WabiSabiErrorCode.InvalidNumberOfPresentedCredentials, 
+					$"{requiredNumberOfPresentations} credential presentations were expected but {presentedCount} were received.");
+			}
+
+			// Don't allow balance to go negative. In case this goes below zero 
+			// then there is a problem somewhere because this should not be possible.
+			if (Balance + registrationRequest.DeltaAmount < Money.Zero)
+			{
+				throw new WabiSabiException(WabiSabiErrorCode.NegativeBalance);
+			}
+
+			// Check all the serial numbers are unique. Note that this is checked separately from
+			// ensuring that they haven't been used before, because even presenting a previously
+			// unused credential more than once in the same request is still a double spend.
+			var rangeProofWidth = registrationRequest.IsNullRequest ? 0 : Constants.RangeProofWidth;
+			var allRangeProofsAreCorrectSize = requested.All(x => x.BitCommitments.Count() == rangeProofWidth);
+			if (!allRangeProofsAreCorrectSize)
+			{
+				throw new WabiSabiException(WabiSabiErrorCode.InvalidBitCommitment);
+			} 
+
+			// Check all the serial numbers are unique.
```

Comment:
```suggestion
			// Check all the serial numbers are unique. Note that this is checked separately from
			// ensuring that they haven't been used before, because even presenting a previously
			// unused credential more than once in the same request is still a double spend.
```